keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ResourceEvent.kt

```diff
@@ -367,18 +376,35 @@ data class ResourceCheckUnresolvable(
     )
 }
 
+enum class ResourceCheckErrorOrigin {
+  @JsonProperty("user") USER,
+  @JsonProperty("system") SYSTEM;
+
+  companion object {
+    fun fromException(exceptionType: Class<out SpinnakerException>) =
+      when {
+        UserException::class.java.isAssignableFrom(exceptionType) -> USER
+        SystemException::class.java.isAssignableFrom(exceptionType) -> SYSTEM
+        else -> throw SystemException("All keel exceptions should inherit from UserException or SystemException. This is a bug.")
+      }
+  }
+}
```

Comment:
Can you explain how this works?

Reply:
So this looks at the `exceptionType` passed to the `ResourceCheckError` constructor to determine a high-level "type" describing where the error originated, which I called `origin`. If it's a sub-class of `UserException`, it's a `USER` error, if it's a sub-class of `SystemException`, it's a `SYSTEM` error.

Reply:
Can the `else` ever happen? It looks like it will always be one or the other, right?

Reply:
It can because the type of `exceptionType` is `Class<SpinnakerException>`, and there are other sub-classes of `SpinnakerException` beyond those two.